[wasm] Enable SIMD

radekdoulik opened the pull request:
Change default value of `WasmEnableSIMD` property to true. That enables SIMD for wasm/AOT.

Use docker image with newer v8 and node on helix.

radekdoulik committed:
[wasm] Enable SIMD

radekdoulik committed:
Let old V8 use simd

Co-authored-by: Ankit Jain <radical@gmail.com>

radekdoulik commented:
Currently we fail on helix because v8 is too old. I am trying to improve that with 2 independent approaches, https://github.com/dotnet/runtime/pull/80808 and https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/774

radical committed:
[wasm] ManagedToNativeGenerator: Skip unmanaged dlls

.. instead crashing with an exception like:

```
src/mono/wasm/build/WasmApp.Native.targets(296,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Build) The "ManagedToNativeGenerator" task failed unexpectedly.
System.BadImageFormatException: This PE image is not a managed executable.
   at System.Reflection.MetadataLoadContext.LoadFromStreamCore(Stream peStream)
   at System.Reflection.MetadataLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at PInvokeTableGenerator.Generate(String[] pinvokeModules, String[] assemblies, String outputPath) in /_/src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:line 42
   at ManagedToNativeGenerator.ExecuteInternal() in /_/src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs:line 68
   at ManagedToNativeGenerator.Execute() in /_/src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs:line 53
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

The wasm targets currently are not able to differentiate the managed
assemblies from the unmanaged ones. so it needs to be handled here.

radical committed:
Merge remote-tracking branch 'origin/main' into pr-wasm-enable-simd

radical committed:
[wasm] WasmApp.Native.targets: do not trigger relinking when WasmEnableSIMD=true

Instead, let the property be effective only for the AOT case.

radekdoulik committed:
Use staging image with newer v8

radical committed:
[wasm] WasmAppHost: Add support for host arguments

radical committed:
[wasm] console template: Add --experimental-wasm-simd to node, and remove v8 config as that isn't supported

radical committed:
[wasm] Pass --experiment-wasm-simd for aot library tests too

radical committed:
Merge branch 'pr-wasm-enable-simd' of https://github.com/radekdoulik/runtime into pr-wasm-enable-simd

radical commented:
/azp run runtime-wasm

radekdoulik committed:
Try to pass full path to v8

Also remove --experimental-wasm-simd option, which shouldn't be needed
for newer v8

radekdoulik committed:
Use docker image for all wasm helix jobs

As we need newer v8

radekdoulik committed:
Set DOTNET_CLI_HOME under workitem payload

The correlation payload is read only

radekdoulik committed:
Set DOTNET_CLI_HOME under workitem payload

The correlation payload is read only. And even with
`export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1` we have still many
failures.

radekdoulik committed:
Revert "Try to pass full path to v8"

This reverts commit a15b3ef5dcab31b0abe2bc914eee6299df7cf56f.

radekdoulik committed:
Use DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1

Let see whether it will avoid all the writes to correlation payload
path

radekdoulik committed:
Set NeedsEMSDKNode to false

To try if we can use newer node in path

radical commented:
[wbt failure]( https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-80797-merge-fd8d36aa7ab4451092/Workloads-Wasm.Build.Tests.NativeBuildTests/1/console.7e7856c0.log?helixlogtype=result):
```
Precompiling failed for /root/helix/work/workitem/e/wbt/3ca4v33t.hq2/obj/Release/net8.0/browser-wasm/wasm/for-publish/aot-in/System.Private.CoreLib.dll with exit code 134. [/root/helix/work/workitem/e/wbt/3ca4v33t.hq2/mono_aot_cross_Release_True.csproj]
Mono Ahead of Time compiler - compiling assembly /root/helix/work/workitem/e/wbt/3ca4v33t.hq2/obj/Release/net8.0/browser-wasm/wasm/for-publish/aot-in/System.Private.CoreLib.dll [/root/helix/work/workitem/e/wbt/3ca4v33t.hq2/mono_aot_cross_Release_True.csproj]
  %6 = load %System.Numerics.Vector4, %System.Numerics.Vector4* %5, align 1 [/root/helix/work/workitem/e/wbt/3ca4v33t.hq2/mono_aot_cross_Release_True.csproj]
<4 x float> [/root/helix/work/workitem/e/wbt/3ca4v33t.hq2/mono_aot_cross_Release_True.csproj]
* Assertion: should not be reached at /__w/1/s/src/mono/mono/mini/mini-llvm.c:1455 [/root/helix/work/workitem/e/wbt/3ca4v33t.hq2/mono_aot_cross_Release_True.csproj]
```

radical commented:
/azp run runtime-wasm

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-enable-simd

radekdoulik committed:
Revert "Set DOTNET_CLI_HOME under workitem payload"

This reverts commit 5dec1aaab403f4cbbd7b0e9f03d965e96efde51c.

radekdoulik committed:
Reverse the test here as we don't relink anymore

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-enable-simd

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-enable-simd

radekdoulik committed:
Set _ExtraTrimmerArgs for tests

So that linker substitution files are used

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
Put the SIMD/trim related props in separate group

More changes from Ankit

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
Add dynamic dependency

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
Disable Vector128IsHardwareAcceleratedTest

Til we have the interpreter support in place

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-enable-simd

radekdoulik committed:
Disable GenericVectorTests.IsHardwareAcceleratedTest

Til we have the interpreter support in place

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
Disable check_no_intrinsic_cattr in emit_vector_2_3_4

And add note about opened issue

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-enable-simd

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
[mono] Disable few quaternion intrinsics

For * and / operators as these are not the same as vector operations.
Opened https://github.com/dotnet/runtime/issues/82408 to implement
correct ones in future.

radekdoulik committed:
Change comparison order

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-enable-simd

radekdoulik commented:
/azp run runtime-wasm

radekdoulik committed:
Use stable images

radekdoulik commented:
/azp run runtime-wasm

radical committed:
Merge branch 'main' into pr-wasm-enable-simd

radical committed:
WBT: Add --engine-arg=--experimental-wasm-simd for node, and v8

lewing reviewed (commented):
LGTM but we need to resolve the wbt failure

radical commented on src/libraries/sendtohelix-wasm.targets:
```diff
@@ -46,7 +46,7 @@
     <NeedsWorkload Condition="'$(Scenario)' == 'BuildWasmApps'">true</NeedsWorkload>
     <NeedsEMSDK Condition="'$(NeedsToBuildWasmAppsOnHelix)' == 'true'">true</NeedsEMSDK>
-    <NeedsEMSDKNode Condition="'$(Scenario)' == 'WasmTestOnNodeJS' or '$(Scenario)' == 'BuildWasmApps'">true</NeedsEMSDKNode>
+    <NeedsEMSDKNode Condition="'$(Scenario)' == 'WasmTestOnNodeJS' or '$(Scenario)' == 'BuildWasmApps'">false</NeedsEMSDKNode>
```
Why this change?

radekdoulik replied:
It comes from https://github.com/dotnet/runtime/commit/37080d6d5d32de323498d60b11ee24e873e56a3b. We use the newer node from the docker image.

radical commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -124,6 +124,11 @@
     <HelixPreCommand Include="printenv | grep COMPlus" />
   </ItemGroup>
 
+  <ItemGroup>
+    <HelixPreCommand Condition="'$(WindowsShell)' == 'true'" Include="set DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1" />
+    <HelixPreCommand Condition="'$(WindowsShell)' != 'true'" Include="export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1" />
+  </ItemGroup>
```
I think this shouldn't be needed because we set `DOTNET_CLI_HOME` to a writeable location now, its line 132-133.

radical commented:
/azp run runtime-wasm

radekdoulik committed:
Do not skip dotnet 1st time experience anymore

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-enable-simd

radekdoulik commented:
/azp run runtime-wasm

radekdoulik commented:
The `runtime-wasm (Build browser-wasm linux Release LibraryTests_NodeJs)` fails on main too.

lewing commented:
> The `runtime-wasm (Build browser-wasm linux Release LibraryTests_NodeJs)` fails on main too.

@maraf are you looking into this?

kg commented on eng/testing/WasmRunnerAOTTemplate.sh:
```diff
@@ -35,8 +35,10 @@ if [[ "$XHARNESS_COMMAND" == "test" ]]; then
+			JS_ENGINE_ARGS="$JS_ENGINE_ARGS --engine-arg=--experimental-wasm-simd"
 		else
 			JS_ENGINE="--engine=V8"
+			JS_ENGINE_ARGS="$JS_ENGINE_ARGS --engine-arg=--experimental-wasm-simd"
```
are we going to keep any lane that tests what happens if simd is disabled? or are we just going to require wasm simd for P3 and later, period. i don't have a good sense of how widespread browsers without wasm simd support are and whether we need to have automated test coverage for what happens

radekdoulik replied:
I will add this test in folloup PR.

radekdoulik replied:
https://github.com/dotnet/runtime/pull/82845

kg commented on src/mono/wasm/templates/templates/console/runtimeconfig.template.json:
```diff
@@ -4,12 +4,10 @@
+                "Host": "nodejs",
+                "host-args": [
+                    "--experimental-wasm-simd"
+                ]
```
why does this diff show v8 being removed?

radekdoulik replied:
@radical is that intentional?

radical replied:
We don't support v8 for the console template, only nodejs.

radekdoulik replied:
so this is unrelated to SIMD, right?

kg reviewed (approved):
Looks good, with the caveat that I don't have the ability to fully vet the yml or msbuild bits (but the latter do look correct to me), and I left some comments
